### PR TITLE
Add support for formatting stdin to stdout

### PR DESCRIPTION
**Pull request by MageJohn**
# Description

Black, along with many other formatters, supports formatting code from stdin, writing it to stdout again. This is particularly useful when integrating the formatter with a code editor, where you often want to format the contents of a buffer without saving the file to disk.

The way Black implements this is through the Unix convention of accepting a '-' as one of the source paths, which is what I've done here.

I reorganised `core.py` a little to make it easier to redirect I/O to streams, rather than always to paths. 

I have also tried to update the tests for the change. I added tests in `test_cli.py`, and while reading the tests and figuring out where best to make additions, I made a few tweaks to some existing tests. Check the individual commits for specifics. Let me know if you'd rather those changes weren't in this PR!

If you're happy with the code changes, I'll go ahead and update the documentation too.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### CODE REVIEW

File: reformat_gherkin/core.py [hunk 1]
```diff
@@ -39,23 +41,48 @@ def find_sources(src: Tuple[str]) -> Set[Path]:
...
+
+    try:
+        return reformat_streams(in_stream, out_stream, options=options)
+    finally:
+        if out_stream is not None:
+            out_stream.close()
+
+
+def reformat_streams(
+    input: BinaryIO, output: Optional[BinaryIO], *, options: Options
```
Comment (ducminh-phan) [inline reformat_gherkin/core.py:83]: `input` is a built-in function, maybe we should have a different name?

**Review by ducminh-phan — COMMENTED** [review]
@MageJohn Thank you for your contribution! I have some suggestions below, could you take a look?

Also, how is a user supposed to use the tool after this change? When I try `cat test.feature | reformat-gherkin -`, it throws an exception
```python
  File "/Users/ducminh/PycharmProjects/reformat-gherkin/reformat_gherkin/cli.py", line 138, in main
    reformat(src, report, options=options)
  File "/Users/ducminh/PycharmProjects/reformat-gherkin/reformat_gherkin/core.py", line 51, in reformat
    changed = reformat_stdin(options=options)
  File "/Users/ducminh/PycharmProjects/reformat-gherkin/reformat_gherkin/core.py", line 64, in reformat_stdin
    return reformat_streams(sys.stdin.buffer, output, options=options)
  File "/Users/ducminh/PycharmProjects/reformat-gherkin/reformat_gherkin/core.py", line 85, in reformat_streams
    src_contents, encoding, existing_newline = decode_stream(input)
  File "/Users/ducminh/PycharmProjects/reformat-gherkin/reformat_gherkin/utils.py", line 78, in decode_stream
    src.seek(0)
io.UnsupportedOperation: File or stream is not seekable.
```

**Comment by MageJohn** [comment]
Thanks for the review! With regards to how it was meant to be used, I've been testing it with `reformat-gherkin - < test.feature`, which was working. I didn't realise that different methods of piping stdin would behave differently! I'll take a look.

File: reformat_gherkin/core.py [hunk 2]
```diff
@@ -39,23 +41,48 @@ def find_sources(src: Tuple[str]) -> Set[Path]:
...
+
+    out_stream = (
+        open(path, "wb") if options.write_back == WriteBackMode.INPLACE else None
+    )
+
+    try:
+        return reformat_streams(in_stream, out_stream, options=options)
+    finally:
+        if out_stream is not None:
+            out_stream.close()
```
[the remark this answers is not in the document]
Comment (MageJohn) [reply reformat_gherkin/core.py:79]: Sounds perfect! I'll make the change.

File: tests/helpers.py [hunk 3]
```diff
@@ -23,6 +23,25 @@ def make_options(
     )
 
 
+def options_to_cli(options):
```
[the remark this answers is not in the document]
Comment (MageJohn) [reply tests/helpers.py:26]: Sure thing, that is a little clearer.

File: reformat_gherkin/core.py (hunk 1 above)
Comment (MageJohn) [reply reformat_gherkin/core.py:83]: Ah, whoops!

**Comment by MageJohn** [comment]
There, those are the three suggestions implemented. I'll dig into the bug deeper later this evening.

**Comment by MageJohn** [comment]
So I ended up rafactoring quite a bit; turns out there was a cleaner way to handle the switching between streams and file paths. It wasn't necessary to solve the bug, but it reduces the number of times the input file is copied about in memory.

**Comment by MageJohn** [comment]
Hey @ducminh-phan! I encountered an issue because my editor was replacing the contents of the file with the contents of stdout even when stdout was empty, so now the tool will always print to stdout even if nothing changed. I also updated the README and bumped the patch version; hope that's all okay.

**Review by ducminh-phan — APPROVED** [review]

**Pull request merged**